addon/-private/system/store.js

```diff
@@ -2417,20 +2419,63 @@ Store = Service.extend({
     heimdall.increment(_setupRelationships);
     let setupToken = heimdall.start('store._setupRelationships');
     let pushed = this._pushedInternalModels;
+    let store = this;
 
     // Cache the inverse maps for each modelClass that we visit during this
     // payload push.  In the common case where we are pushing many more
     // instances than types we want to minimize the cost of looking up the
     // inverse map and the overhead of Ember.get adds up.
     let modelNameToInverseMap = Object.create(null);
+    let modelNameToRelationshipsMap = Object.create(null);
 
-    for (let i = 0, l = pushed.length; i < l; i += 2) {
-      // This will convert relationships specified as IDs into DS.Model instances
-      // (possibly unloaded) and also create the data structures used to track
-      // relationships.
+    for (let i = 0; i < pushed.length; i += 2) {
```

Comment:
this produces quite sketchy ES5 output, lots of nested functions to handle the `let` TDZ

```js
  _setupRelationships: function _setupRelationships() {
    var pushed = this._pushedInternalModels;
    var store = this;

    // Cache the inverse maps for each modelClass that we visit during this
    // payload push.  In the common case where we are pushing many more
    // instances than types we want to minimize the cost of looking up the
    // inverse map and the overhead of Ember.get adds up.
    var modelNameToInverseMap = Object.create(null);
    var modelNameToRelationshipsMap = Object.create(null);

    var _loop2 = function _loop2(i) {
      var internalModel = pushed[i];
      var data = pushed[i + 1];
      var modelName = internalModel.modelName;
      var relationshipsByName = modelNameToRelationshipsMap[modelName];

      if (relationshipsByName === undefined) {
        relationshipsByName = modelNameToRelationshipsMap[modelName] = get$3(internalModel.modelClass, 'relationshipsByName');
      }

      var relationshipNames = relationshipsByName._keys.list;
      var relationships = internalModel._relationships;

      var _loop3 = function _loop3(keyIndex) {
        var relationshipName = relationshipNames[keyIndex];
        var relationshipData = data.relationships[relationshipName];

        if (relationshipData !== undefined) {
          var relationshipRequiresNotification = relationships.has(relationshipName) || isInverseRelationshipInitialized(store, internalModel, data, relationshipName, modelNameToInverseMap);

          if (relationshipRequiresNotification) {
            relationships.get(relationshipName).push(relationshipData);
          }

          // in debug, assert payload validity eagerly
        }
      };

      for (var keyIndex = 0; keyIndex < relationshipNames.length; keyIndex++) {
        _loop3(keyIndex);
      }
    };

    for (var i = 0; i < pushed.length; i += 2) {
      _loop2(i);
    }

    pushed.length = 0;
  }
```